### PR TITLE
[Diffusion][LoRA] Rework adapters and implement LoKr adapter

### PR DESCRIPTION
**Pull request by vlserov**
Implementation LoKr adapter base on Roadmap issue #25897

## Motivation

LoKr (Low-Rank Kronecker Product) is a parameter-efficient fine-tuning method and a variant of the popular LoRA technique. LoKr approximates a large weight update matrix by decomposing it into two low-rank matrices and combining them using the **Kronecker product**. 

llustration-of-the-tensor-Kronecker-Product-A-B-C-Right-A-2-2-2-tensor-B-and-a

LoKr introduces more expressive adaptation layers by utilising Kronecker products, allowing the model to capture more complex interactions and patterns within the data without significantly increasing the number of parameters.
This boosts expressive power, enabling the model to capture complex data patterns without a heavy parameter increase.

For Stable Diffusion, LoKr offers clear benefits:

- Superior detail by modeling complex spatial textures for more realistic outputs.
- Better style transfer, capturing aesthetic nuances for faithful artistic adaptation.
- Efficient adaptation, achieving strong performance without full parameter updates, saving resources.
- Flexible application, as it can be deployed across various model layers for task-specific tuning.

## Modifications

1. Rework sglang.multimodal_gen.runtime.layers module: generalise behaviour for parallel execution for case inherited from LinearBase, and create ParallelLinear, which implements such behaviour
2. Implementation LoKr based on diagram in  #25897

```mermaid
classDiagram

class BaseLayerWithLoRA {
+List~BaseWeightEntry~ required_adapter_type
+forward(Tensor)
}

class Adapter {
+delta() Tensor
+iterate(Tensor) Tensor
+split_weights(rank, size, layout) Adapter
}

class LoRAAdapter {
+delta() Tensor
+iterate(Tensor) Tensor
+split_weights(rank, size, layout) LoRAAdapter
}

class LoKrAdapter {
+delta() Tensor
+iterate(Tensor) Tensor
+split_weights(rank, size, layout) LoKrAdapter
}

class BaseWeightEntry {
+register()
}

class LoRAWeightEntry {
+createLoRAWeight() LoRAWeightEntry
}

class LoKrWeightEntry {
+createLoKrWeight() LoKrWeightEntry
}

Adapter <|-- LoRAAdapter
Adapter <|-- LoKrAdapter
BaseWeightEntry <|-- LoRAWeightEntry
BaseWeightEntry <|-- LoKrWeightEntry
BaseWeightEntry  o-- BaseLayerWithLoRA
LoRAAdapter  <-- LoRAWeightEntry
LoKrAdapter  <-- LoKrWeightEntry
```

## Accuracy Tests

For test **Tongyi-MAI/Z-Image-Turbo** has been used as a base model, **Ttio2/Z-Image-Turbo-pencil-sketch** as a LoRA adapter, **F16/z-image-turbo-flow-dpo** as LoKr adapter.

| Model | Peak Memory in MB | Inference Time in sec |
|--------|--------|--------|
| Base Model | 15342 | 3.3874373009894043 |
| Base Model with LoRA | 16180 | 3.44018611201318 |
| Base Model with LoKr | 16216 | 3.44269112899201 |

<details>
<summary>Command line for server launch and request</summary>

Server launch
`HCCL_BUFFSIZE=256 sglang serve --model-path Tongyi-MAI/Z-Image-Turbo --model-type diffusion --dit-cpu-offload false   --num-gpus 1 --port <port number>`

Request 
`curl -sS -X POST "http://localhost:<port number>/v1/images/generations" -H "Content-Type: application/json"  -H "Authorization: Bearer sk-proj-1234567890"  -d '{ "prompt": "woman, Asian ethnicity, white dress, looking away, long hair, outdoor setting, building facade, plants, serene expression, elegance, side profile, standing, daylight, soft focus, pastel colors, fashion, youthful, casual elegance, architectural elements, natural light, tassel detail on dress",  "size": "1024x1024", "response_format": "b64_json"}'`

</details>

<details>
<summary>Command line for setup adapter</summary>

For LoRA adapter:
```
curl -X POST http://localhost:<port number>/v1/set_lora -H "Content-Type: application/json" -d '{"lora_nickname": "lora_name", "lora_path": "Ttio2/Z-Image-Turbo-pencil-sketch", "target": "all", "strength": 1}'
curl -X POST http://localhost:<port number>/v1/merge_lora_weights -H "Content-Type: application/json" -d '{"strength": 1.0}'
```

For LoKr adapter:
```
curl -X POST http://localhost:<port number>/v1/set_lora -H "Content-Type: application/json" -d '{"lora_nickname": "lora_name", "lora_path": "F16/z-image-turbo-flow-dpo", "target": "all", "strength": 1}'
curl -X POST http://localhost:<port number>/v1/merge_lora_weights -H "Content-Type: application/json" -d '{"strength": 1.0}'
```
</details>

<details>
<summary>Output images:</summary>

Base Model
*Base Model*
LoRA Adapter
*Base Model with LoRA Adapter*
LoKr Adapter
*Base Model with LoKr Adapter*

### Closer look for images with different adapters
Adapter **ifmylove2011/girlslike-zimage** has been trained for drawing different faces. Example below has been shown how woman face is changed depending on used adapter.
LoKr Adapter
*Base Model with different LoKr Adapters*

</details>

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
4. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
5. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
6. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** -- add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.
Latest PR Test (AMD ROCm 7.2): :heavy_minus_sign: **No AMD PR run found for this commit**.

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/layers/lora/linear.py [hunk 1]
```diff
@@ -34,45 +36,412 @@
...
+
+
+class ParallelLayout(Enum):
+    NoneParallelLayout = 0
+    RowwiseParallelLayout = 1
+    ColwiseParallelLayout = 2
+
+
+class Adapter:
+    def to_local(weights: dict[str, torch.Tensor]):
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/layers/lora/linear.py:78]: this function is neither `self` nor `staticmethod`, please double check it

File: python/sglang/multimodal_gen/runtime/layers/lora/linear.py [hunk 2]
```diff
@@ -34,45 +36,412 @@
...
+        def _to_local(tensor: torch.Tensor):
+            return tensor.to_local() if isinstance(tensor, DTensor) else tensor
+
+        return _recursive_apply(_to_local, weights)
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/layers/lora/linear.py:82]: i don't think we need `_recursive_apply` here, you just need to reuse `named_parameters` from PyTorch

```python
for name, param in self.named_parameters(recurse=False):
    if isinstance(param.data, DTensor):
        setattr(self, name, nn.Parameter(param.data.to_local()))
```

File: python/sglang/multimodal_gen/runtime/layers/lora/linear.py [hunk 3]
```diff
@@ -34,45 +36,412 @@
...
+    @property
+    def dtype(self) -> torch.dtype | None:
+        any_item = next(_recursive_by_type(self.weights, torch.Tensor), None)
+        return any_item.dtype if any_item is not None else None
+
+    def to(self, *args, **kwargs):
+        def _tensor_to(tensor: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+            return tensor.to(*args, **kwargs)
+
+        return _recursive_apply(_tensor_to, self.weights, *args, **kwargs)
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/layers/lora/linear.py:93]: same problems as before

File: python/sglang/multimodal_gen/runtime/layers/lora/linear.py [hunk 4]
```diff
@@ -34,45 +36,412 @@
...
+            _recursive_apply(_tensor_to, self.lora_B, *args, **kwargs),
+        )
+
+    def split_weights(
+        self,
+        distributed_rank: int,
+        distributed_size: int,
+        layout: ParallelLayout,
+        base_layer,
+    ) -> "LoRAAdapter":
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/layers/lora/linear.py:138]: there is no `base_layer` in Base class

File: python/sglang/multimodal_gen/runtime/layers/lora/linear.py [hunk 5]
```diff
@@ -34,45 +36,412 @@
...
+            yield from _recursive_by_type(item, target_type)
+
+
+class ParallelLayout(Enum):
+    NoneParallelLayout = 0
+    RowwiseParallelLayout = 1
+    ColwiseParallelLayout = 2
+
+
+class Adapter:
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/layers/lora/linear.py:77]: you are supposed to use `abc.ABC` here, then `@abstractmethod` will make affect

File: python/sglang/multimodal_gen/runtime/layers/linear.py [hunk 6]
```diff
@@ -220,7 +220,46 @@ def forward(self, x: torch.Tensor) -> tuple[torch.Tensor, Parameter | None]:
         raise NotImplementedError
 
 
-class ReplicatedLinear(LinearBase):
+class ParallelLinear(LinearBase):
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/layers/linear.py:223]: can you explain why you add one more class here

File: python/sglang/multimodal_gen/runtime/pipelines_core/lora_format_adapter.py [hunk 7]
```diff
@@ -11,15 +11,17 @@
 
 
 class LoRAFormat(str, Enum):
-    """Supported external LoRA formats before normalization."""
+    """Supported external adapter formats before normalization."""
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/pipelines_core/lora_format_adapter.py:14]: please update the class name

File: python/sglang/multimodal_gen/runtime/pipelines_core/lora_format_adapter.py [hunk 8]
```diff
@@ -11,15 +11,17 @@
...
-    AI_TOOLKIT_FLUX = "ai-toolkit-flux"
+    UNDEFINED = "undefined"
+    LORA_STANDARD = "lora.standard"
+    LORA_NON_DIFFUSERS_SD = "lora.non-diffusers-sd"
+    LORA_QWEN_IMAGE_STANDARD = "lora.qwen-image-standard"
+    LORA_XLABS_FLUX = "lora.xlabs-ai"
+    LORA_KOHYA_FLUX = "lora.kohya-flux"
+    LORA_WAN = "lora.wan"
+    LORA_AI_TOOLKIT_FLUX = "lora.ai-toolkit-flux"
+    LOKR_STANDARD = "lokr.standard "
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/pipelines_core/lora_format_adapter.py:24]: how about we add one more `LoKrFormat` rather than change the exist class?

**Comment by ping1jing2** [comment]
there are more refactor works in your PR, please add more descriptions for this part

File: python/sglang/multimodal_gen/runtime/layers/linear.py (hunk 6 above)
Comment (vlserov) [reply python/sglang/multimodal_gen/runtime/layers/linear.py:223]: Made separate PR #37460 for explaining reason for changes
